Adding log to print all configuration flags at start of gcs fuse mount

ankitaluthra1 opened the pull request:
### Description
Uses flagStorage struct to get all flags which are being used in gcsfuse and add a info log while starting gcsfuse

### Link to the issue in case of a bug fix.
[issue-link](https://b.corp.google.com/issues/315990701)

### Testing details
1. Manual - Done on Local
2. Unit tests - Added unit test for logic of aggregated log message containing all flags
3. Integration tests - NA

raj-prince commented:
1. In this PR, we are only logging passed flags, right? We also need to print the config flags.
2. Could you please update the PR description with the newly added logs?

ankitaluthra1 commented:
Sample log (For now I have kept "" in the log as I am not sure removing " from logs will result in any formatting errors for any customers, let me know if we need to remove "" from log statement)
`{"time":{"timestampSeconds":1703683455,"timestampNanos":1703683455611934894},"severity":"INFO","msg":"Loading config for mount flags {\"AppName\":\"\",\"Foreground\":true,\"ConfigFile\":\"\",\"MountOptions\":{},\"DirMode\":493,\"FileMode\":420,\"Uid\":-1,\"Gid\":-1,\"ImplicitDirs\":false,\"OnlyDir\":\"\",\"RenameDirLimit\":0,\"CustomEndpoint\":null,\"BillingProject\":\"\",\"KeyFile\":\"\",\"TokenUrl\":\"\",\"ReuseTokenFromUrl\":true,\"EgressBandwidthLimitBytesPerSecond\":-1,\"OpRateLimitHz\":-1,\"SequentialReadSizeMb\":200,\"MaxRetrySleep\":30000000000,\"StatCacheCapacity\":4096,\"StatCacheTTL\":60000000000,\"TypeCacheTTL\":60000000000,\"HttpClientTimeout\":0,\"MaxRetryDuration\":-1000000000,\"RetryMultiplier\":2,\"LocalFileCache\":false,\"TempDir\":\"\",\"ClientProtocol\":\"http1\",\"MaxConnsPerHost\":100,\"MaxIdleConnsPerHost\":100,\"EnableNonexistentTypeCache\":false,\"StackdriverExportInterval\":0,\"OtelCollectorAddress\":\"\",\"LogFile\":\"\",\"LogFormat\":\"json\",\"ExperimentalEnableJsonRead\":false,\"DebugFuseErrors\":true,\"DebugFuse\":false,\"DebugFS\":false,\"DebugGCS\":false,\"DebugHTTP\":false,\"DebugInvariants\":false,\"DebugMutex\":false}"}
{"time":{"timestampSeconds":1703683455,"timestampNanos":1703683455611975904},"severity":"INFO","msg":"Loading config from config-file {\"CreateEmptyFile\":false,\"Severity\":\"INFO\",\"Format\":\"json\",\"FilePath\":\"\",\"LogRotateConfig\":{\"MaxFileSizeMB\":512,\"BackupFileCount\":10,\"Compress\":true}}"}`

ankitaluthra1 commented:
> 1. In this PR, we are only logging passed flags, right? We also need to print the config flags.
> 2. Could you please update the PR description with the newly added logs?
> 3. Lint is failing.

Added config file flag as well
Added sample logs in comment
Lint is fixed

raj-prince commented on main.go:
```diff
@@ -224,6 +234,8 @@ func runCLIApp(c *cli.Context) (err error) {
 	}
 
 	logger.Infof("Start gcsfuse/%s for app %q using mount point: %s\n", getVersion(), flags.AppName, mountPoint)
+	logger.Infof("Loading config for mount flags %s", stringify(*flags))
```
nit: instead we can write better like, 1GCSFuse mount command flags: %s`

ankitaluthra1 replied:
Updated logs with above text

raj-prince commented on main_test.go:
```diff
@@ -72,3 +75,64 @@ func (t *MainTest) TestGetUserAgentWhenMetadataImageTypeEnvVarSetAndAppNameNotSe
+	actual := stringify(mountConfig)
+
+	expected := "{\"CreateEmptyFile\":false,\"Severity\":\"TRACE\",\"Format\":\"\",\"FilePath\":\"\\\"path\\\"to\\\"file\\\"\",\"LogRotateConfig\":{\"MaxFileSizeMB\":2,\"BackupFileCount\":2,\"Compress\":true}}"
+	AssertEq(strings.TrimSpace(expected), strings.TrimSpace(actual))
```
Just curious, why do we need TrimSpace here?

ankitaluthra1 replied:
its not needed now actually, have removed it (added it earlier was testing random data without quotes)

raj-prince commented on main.go:
```diff
@@ -185,6 +186,15 @@ func populateArgs(c *cli.Context) (
+	inputBytes, err := json.Marshal(input)
+
+	if err != nil {
+		logger.Errorf("Error in stringify %v", err)
```
We should log it as Warning, if it's not a fatal error.

ankitaluthra1 replied:
done.. changed it to warn

raj-prince commented on main.go:
```diff
@@ -185,6 +186,15 @@ func populateArgs(c *cli.Context) (
 	return
 }
 
+func stringify(input any) string {
```
Can we move this method in util?
internal/util/util.go

ankitaluthra1 replied:
it was only used in one file hence kept it here for now, will it be okay if we move it to utility when more than one file uses it

ankitaluthra1 replied:
additionally, I won't be able to test flagStorage example as flagStorage is private to main package

ankitaluthra1 replied:
have moved Stringify to util package

raj-prince commented on main_test.go:
```diff
@@ -72,3 +75,64 @@ func (t *MainTest) TestGetUserAgentWhenMetadataImageTypeEnvVarSetAndAppNameNotSe
+	AssertEq(strings.TrimSpace(expected), strings.TrimSpace(actual))
+}
+
+type CustomType struct {
```
nit: Is it mandatory to make this object public? Can't we keep `customType`?

ankitaluthra1 replied:
made it private

raj-prince commented on main_test.go:
```diff
@@ -72,3 +75,64 @@ func (t *MainTest) TestGetUserAgentWhenMetadataImageTypeEnvVarSetAndAppNameNotSe
+}
+
+func (t *MainTest) TestStringifyShouldReturnEmptyStringWhenMarshalErrorsOut() {
+
```
nit: extra new line.

ankitaluthra1 replied:
removed

raj-prince reviewed (changes requested):
Few minor comments.

raj-prince commented on internal/util/util_test.go:
```diff
@@ -157,3 +159,53 @@ func (t *UtilTest) TestResolveConfigFilePaths() {
+		"2": 2,
+		"3": 3,
+	}
+
```
nit: Please remove new line, in unit test we use newline to separate arrange, act and assert statment.

ankitaluthra1 replied:
done

raj-prince commented on internal/util/util.go:
```diff
@@ -76,3 +77,13 @@ func ResolveConfigFilePaths(config *config.MountConfig) (err error) {
 	return
 }
+
+func Stringify(input any) string {
```
nit: We can write a doc here, which helps in understanding the method just hovering over the usage.

ankitaluthra1 replied:
done

raj-prince commented on internal/util/util_test.go:
```diff
@@ -157,3 +159,53 @@ func (t *UtilTest) TestResolveConfigFilePaths() {
+}
+
+type customTypeForSuccess struct {
+	Value       string
```
Just curious, what happens if we make the attribute private?

ankitaluthra1 replied:
it was unable to marshal.. I verified

raj-prince commented on internal/util/util_test.go:
```diff
@@ -157,3 +159,53 @@ func (t *UtilTest) TestResolveConfigFilePaths() {
+	actual := Stringify(customInstance)
+
+	expected := ""
+	AssertEq(strings.TrimSpace(expected), strings.TrimSpace(actual))
```
Why TrimSpace is required here?

ankitaluthra1 replied:
its typo... not required, have removed it

raj-prince reviewed (changes requested):
LGTM, Added few nits.

raj-prince commented on internal/util/util.go:
```diff
@@ -76,3 +77,14 @@ func ResolveConfigFilePaths(config *config.MountConfig) (err error) {
 	return
 }
+
+// Returns empty String in case of failure in marshalling the passed input.
```
nit: Stringify marshals an object (only exported attribute) to a JSON string. If marshalling fails, it returns an empty string.

ankitaluthra1 replied:
done

raj-prince reviewed (approved):
LGTM, minor doc change.
